src/GafferScene/Unencapsulate.cpp

```diff
@@ -51,18 +49,90 @@ using namespace GafferScene;
 namespace
 {
 
-// TODO - is there a centralized version of this?
 ScenePlug::ScenePath concatScenePath( const ScenePlug::ScenePath &a, const ScenePlug::ScenePath &b )
 {
-	ScenePlug::ScenePath result = a;
-	result.reserve( result.size() + b.size() );
-	for( const std::string & n : b )
-	{
-		result.push_back( n );
-	}
+	ScenePlug::ScenePath result;
+	result.reserve( a.size() + b.size() );
+	result.insert( result.end(), a.begin(), a.end() );
+	result.insert( result.end(), b.begin(), b.end() );
 	return result;
 }
 
+class CapsuleScope : boost::noncopyable
+{
+
+	public :
+
+		CapsuleScope(
+			const Gaffer::Context *context, const ScenePlug *inPlug,
+			const ScenePlug::ScenePath &parentPath, boost::optional<const ScenePlug::ScenePath &> branchPath, bool requireCapsule, boost::optional<const InternedString &> setName = boost::optional<const InternedString &>()
+		)
+		{
+			m_object = inPlug->object( parentPath );
+			m_capsule = IECore::runTimeCast< const Capsule >( m_object.get() );
+			if( !m_capsule )
+			{
+				if( requireCapsule )
+				{
+					// Note : John requested this exception, but I've been unable to actually devise a test
+					// that would actually trigger it, because at the point where we call CapsuleScope, we
+					// have parentPath and branchPath from parentAndBranchPaths, and parentAndBranchPaths
+					// does its own checking
+					std::string branchPathStr;
+					std::string parentPathStr;
+					ScenePlug::pathToString( *branchPath, branchPathStr );
+					ScenePlug::pathToString( parentPath, parentPathStr );
+					throw IECore::Exception( "Unencapsulating path " + branchPathStr + " within " + parentPathStr + " which is not a capsule." );
+				}
+			}
+			else
+			{
+				m_scope.emplace( m_capsule->context() );
+				if( setName )
+				{
+					m_scope->remove( ScenePlug::scenePathContextName );
```

Comment:
Maybe Encapsulate should be removing this at the point of creation?

Reply:
I had wondered about this myself, hadn't done it just because I wanted to make minimal changes, but it does seem cleaner, so I've done this now.  Note that I've done this change inside Capsule ... in some ways, it would be more logical in Encapsulate, but the construction of a new context that can be edited only happens inside the Capsule constructor.

Reply:
Hmm, working on encapsulation in the Instancer now, we have a need there as well to edit the context of the capsule when we create it.  Maybe the m_context of Capsule shouldn't be const?  It should be treated as const once the Capsule is returned - but at that point, the whole Capsule should be treated at const anyway?

Reply:
Wait, I've got a better idea:  the context parameter to the Capsule constructor should be an r-value reference.  Then the caller is required to construct a new context for the Capsule, and the capsule can safely take ownership of it without needing to copy it.

Reply:
Let's chat about this one in our call tomorrow. I think I did something similar to what you propose in `IECoreScene::ShaderNetwork`, but I still haven't quite got my head round all the rvalue-reference stuff yet.